<a href="https://colab.research.google.com/github/Nhung-Nguyen86/DataScience/blob/main/Spam_filtering_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# SPAM FILTERING DATA

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB

1. Begin by reading the article titled "Spam Filtering with Naive Bayes – Which Naive Bayes?" by Metsis et al., which has been uploaded to Canvas. This article provides insights into the experimental procedure for evaluating Naive Bayes classifiers for spam filtering. Pay particular attention to their approach for comparing different versions of Naive Bayes using ROC curves.

In [ ]:
data = pd.read_csv('Spam Filtering Data.csv')
data

Unnamed: 0                                               Body  Label
0              0  \nSave up to 70% on Life Insurance.\nWhy Spend...      1
1              1  1) Fight The Risk of Cancer!\nhttp://www.adcli...      1
2              2  1) Fight The Risk of Cancer!\nhttp://www.adcli...      1
3              3  ##############################################...      1
4              4  I thought you might like these:\n1) Slim Down ...      1
...          ...                                                ...    ...
6041        6041                                              empty      0
6042        6042                    ___           ___           ...      0
6043        6043  IN THIS ISSUE:01. Readers write\n02. Extension...      0
6044        6044                                              empty      0
6045        6045                                              empty      0

[6046 rows x 3 columns]

In [ ]:
data.head()

Unnamed: 0                                               Body  Label
0           0  \nSave up to 70% on Life Insurance.\nWhy Spend...      1
1           1  1) Fight The Risk of Cancer!\nhttp://www.adcli...      1
2           2  1) Fight The Risk of Cancer!\nhttp://www.adcli...      1
3           3  ##############################################...      1
4           4  I thought you might like these:\n1) Slim Down ...      1

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6046 entries, 0 to 6045
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  6046 non-null   int64 
 1   Body        6045 non-null   object
 2   Label       6046 non-null   int64 
dtypes: int64(2), object(1)
memory usage: 141.8+ KB


In [ ]:
 data.describe()

Unnamed: 0        Label
count  6046.000000  6046.000000
mean   3022.500000     0.313596
std    1745.474195     0.463993
min       0.000000     0.000000
25%    1511.250000     0.000000
50%    3022.500000     0.000000
75%    4533.750000     1.000000
max    6045.000000     1.000000

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split

data = data.drop(columns=['Unnamed: 0'])

data = data.dropna(subset=['Body'])

tfidf = TfidfVectorizer(stop_words='english', max_features=1000)

X = tfidf.fit_transform(data['Body'])

y = data['Label']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

X_train.shape, X_test.shape, y_train.shape, y_test.shape


((4836, 1000), (1209, 1000), (4836,), (1209,))

In [ ]:
data.isnull().sum()

Body     0
Label    0
dtype: int64

2. Implement the following machine learning classifiers:
Naive Bayes Classifier (Following the approaches discussed in the provided article)
Support Vector Machines (SVM)
Logistic Regression
Gradient Boosting Machines (GBM)
Random Forest Classifier

# Naive Bayes Classifier - MultinomialNB

In [ ]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_curve, auc, confusion_matrix, classification_report
import matplotlib.pyplot as plt

naive_bayes_classifier = MultinomialNB()
naive_bayes_classifier.fit(X_train, y_train)

nb_predictions = naive_bayes_classifier.predict(X_test)
nb_probabilities = naive_bayes_classifier.predict_proba(X_test)[:, 1]  # Probabilities for the positive class

nb_accuracy = accuracy_score(y_test, nb_predictions)
nb_precision = precision_score(y_test, nb_predictions)
nb_recall = recall_score(y_test, nb_predictions)
nb_f1_score = f1_score(y_test, nb_predictions)
nb_confusion = confusion_matrix(y_test, nb_predictions)
nb_classification_report = classification_report(y_test, nb_predictions)

print("Accuracy:", nb_accuracy)
print("Precision:", nb_precision)
print("Recall:", nb_recall)
print("F1 Score:", nb_f1_score)
print("Confusion Matrix:\n", nb_confusion)
print("Classification Report:\n", nb_classification_report)


Accuracy: 0.9106699751861043
Precision: 0.9537037037037037
Recall: 0.7686567164179104
F1 Score: 0.8512396694214877
Confusion Matrix:
 [[792  15]
 [ 93 309]]
Classification Report:
               precision    recall  f1-score   support

           0       0.89      0.98      0.94       807
           1       0.95      0.77      0.85       402

    accuracy                           0.91      1209
   macro avg       0.92      0.88      0.89      1209
weighted avg       0.91      0.91      0.91      1209



# Naive Bayes Classifier - GaussianNB

In [ ]:
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_curve, auc, confusion_matrix, classification_report

X_train_dense = X_train.toarray()
X_test_dense = X_test.toarray()

gaussian_nb_classifier = GaussianNB()
gaussian_nb_classifier.fit(X_train_dense, y_train)

nb_predictions = gaussian_nb_classifier.predict(X_test_dense)
nb_probabilities = gaussian_nb_classifier.predict_proba(X_test_dense)[:, 1]

nb_accuracy = accuracy_score(y_test, nb_predictions)
nb_precision = precision_score(y_test, nb_predictions)
nb_recall = recall_score(y_test, nb_predictions)
nb_f1_score = f1_score(y_test, nb_predictions)
nb_confusion = confusion_matrix(y_test, nb_predictions)
nb_classification_report = classification_report(y_test, nb_predictions)

fpr, tpr, thresholds = roc_curve(y_test, nb_probabilities)
nb_auc = auc(fpr, tpr)

print("Accuracy:", nb_accuracy)
print("Precision:", nb_precision)
print("Recall:", nb_recall)
print("F1 Score:", nb_f1_score)
print("Confusion Matrix:\n", nb_confusion)
print("Classification Report:\n", nb_classification_report)


Accuracy: 0.9363110008271298
Precision: 0.8684807256235828
Recall: 0.9527363184079602
F1 Score: 0.9086595492289442
Confusion Matrix:
 [[749  58]
 [ 19 383]]
Classification Report:
               precision    recall  f1-score   support

           0       0.98      0.93      0.95       807
           1       0.87      0.95      0.91       402

    accuracy                           0.94      1209
   macro avg       0.92      0.94      0.93      1209
weighted avg       0.94      0.94      0.94      1209



# Naive Bayes Classifier - BernoulliNB

In [ ]:
from sklearn.naive_bayes import BernoulliNB
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_curve, auc, confusion_matrix, classification_report
import matplotlib.pyplot as plt

bernoulli_nb_classifier = BernoulliNB()
bernoulli_nb_classifier.fit(X_train, y_train)

nb_predictions = bernoulli_nb_classifier.predict(X_test)
nb_probabilities = bernoulli_nb_classifier.predict_proba(X_test)[:, 1]  # Probabilities for the positive class

nb_accuracy = accuracy_score(y_test, nb_predictions)
nb_precision = precision_score(y_test, nb_predictions)
nb_recall = recall_score(y_test, nb_predictions)
nb_f1_score = f1_score(y_test, nb_predictions)
nb_confusion = confusion_matrix(y_test, nb_predictions)
nb_classification_report = classification_report(y_test, nb_predictions)

fpr, tpr, thresholds = roc_curve(y_test, nb_probabilities)
nb_auc = auc(fpr, tpr)

print("Accuracy:", nb_accuracy)
print("Precision:", nb_precision)
print("Recall:", nb_recall)
print("F1 Score:", nb_f1_score)
print("Confusion Matrix:\n", nb_confusion)
print("Classification Report:\n", nb_classification_report)


Accuracy: 0.9396195202646815
Precision: 0.8764302059496567
Recall: 0.9527363184079602
F1 Score: 0.9129916567342075
Confusion Matrix:
 [[753  54]
 [ 19 383]]
Classification Report:
               precision    recall  f1-score   support

           0       0.98      0.93      0.95       807
           1       0.88      0.95      0.91       402

    accuracy                           0.94      1209
   macro avg       0.93      0.94      0.93      1209
weighted avg       0.94      0.94      0.94      1209



#  Support Vector Machine (SVM)

In [ ]:
from sklearn.svm import SVC

svm_classifier = SVC(kernel='linear')
svm_classifier.fit(X_train, y_train)

svm_predictions = svm_classifier.predict(X_test)

svm_accuracy = accuracy_score(y_test, svm_predictions)
svm_precision = precision_score(y_test, svm_predictions)
svm_recall = recall_score(y_test, svm_predictions)
svm_f1_score = f1_score(y_test, svm_predictions)
svm_confusion = confusion_matrix(y_test, svm_predictions)
svm_classification_report = classification_report(y_test, svm_predictions)

svm_auc = auc(fpr, tpr)

print("Accuracy:", svm_accuracy)
print("Precision:", svm_precision)
print("Recall:", svm_recall)
print("F1 Score:", svm_f1_score)
print("Confusion Matrix:\n", svm_confusion)
print("Classification Report:\n", svm_classification_report)

Accuracy: 0.9511993382961125
Precision: 0.8853932584269663
Recall: 0.9800995024875622
F1 Score: 0.9303423848878395
Confusion Matrix:
 [[756  51]
 [  8 394]]
Classification Report:
               precision    recall  f1-score   support

           0       0.99      0.94      0.96       807
           1       0.89      0.98      0.93       402

    accuracy                           0.95      1209
   macro avg       0.94      0.96      0.95      1209
weighted avg       0.95      0.95      0.95      1209



The Support Vector Machine (SVM) Classifier performed even better, achieving an accuracy of approximately 95.12% on the test data.

# Logistic Regression classifier

In [ ]:
from sklearn.linear_model import LogisticRegression

logistic_classifier = LogisticRegression(max_iter=1000)
logistic_classifier.fit(X_train, y_train)

logistic_predictions = logistic_classifier.predict(X_test)

logistic_accuracy = accuracy_score(y_test, logistic_predictions)
logistic_classification_report = classification_report(y_test, logistic_predictions)

logistic_accuracy, logistic_classification_report

logistic_accuracy = accuracy_score(y_test, logistic_predictions)
logistic_precision = precision_score(y_test, logistic_predictions)
logistic_recall = recall_score(y_test, logistic_predictions)
logistic_f1_score = f1_score(y_test, logistic_predictions)
logistic_confusion = confusion_matrix(y_test, logistic_predictions)
logistic_classification_report = classification_report(y_test, logistic_predictions)


logistic_auc = auc(fpr, tpr)

print("Accuracy:", logistic_accuracy)
print("Precision:", logistic_precision)
print("Recall:", logistic_recall)
print("F1 Score:", logistic_f1_score)
print("Confusion Matrix:\n", logistic_confusion)
print("Classification Report:\n", logistic_classification_report)


Accuracy: 0.9528535980148883
Precision: 0.9020979020979021
Recall: 0.9626865671641791
F1 Score: 0.9314079422382671
Confusion Matrix:
 [[765  42]
 [ 15 387]]
Classification Report:
               precision    recall  f1-score   support

           0       0.98      0.95      0.96       807
           1       0.90      0.96      0.93       402

    accuracy                           0.95      1209
   macro avg       0.94      0.96      0.95      1209
weighted avg       0.95      0.95      0.95      1209



#The Logistic Regression classifier achieved an accuracy of approximately 95.29% on the test data

# Gradient Boosting Classifier

In [ ]:
from sklearn.ensemble import GradientBoostingClassifier

gbm_classifier = GradientBoostingClassifier()
gbm_classifier.fit(X_train, y_train)

gbm_predictions = gbm_classifier.predict(X_test)

gbm_accuracy = accuracy_score(y_test, gbm_predictions)
gbm_precision = precision_score(y_test, gbm_predictions)
gbm_recall = recall_score(y_test, gbm_predictions)
gbm_f1_score = f1_score(y_test, gbm_predictions)
gbm_confusion = confusion_matrix(y_test, gbm_predictions)
gbm_classification_report = classification_report(y_test, gbm_predictions)

print("Accuracy:", gbm_accuracy)
print("Precision:", gbm_precision)
print("Recall:", gbm_recall)
print("F1 Score:", gbm_f1_score)
print("Confusion Matrix:\n", gbm_confusion)
print("Classification Report:\n", gbm_classification_report)


Accuracy: 0.9429280397022333
Precision: 0.8899297423887588
Recall: 0.945273631840796
F1 Score: 0.916767189384801
Confusion Matrix:
 [[760  47]
 [ 22 380]]
Classification Report:
               precision    recall  f1-score   support

           0       0.97      0.94      0.96       807
           1       0.89      0.95      0.92       402

    accuracy                           0.94      1209
   macro avg       0.93      0.94      0.94      1209
weighted avg       0.94      0.94      0.94      1209



#The Gradient Boosting Machine (GBM) classifier achieved an accuracy of approximately 94.21% on the test data.

# Random Forest classifier

In [ ]:
from sklearn.ensemble import RandomForestClassifier

random_forest_classifier = RandomForestClassifier()
random_forest_classifier.fit(X_train, y_train)

rf_predictions = random_forest_classifier.predict(X_test)

rf_accuracy = accuracy_score(y_test, rf_predictions)
rf_classification_report = classification_report(y_test, rf_predictions)

rf_accuracy, rf_classification_report

rf_accuracy = accuracy_score(y_test, rf_predictions)
rf_precision = precision_score(y_test, rf_predictions)
rf_recall = recall_score(y_test, rf_predictions)
rf_f1_score = f1_score(y_test, rf_predictions)
rf_confusion = confusion_matrix(y_test, rf_predictions)
rf_classification_report = classification_report(y_test, rf_predictions)


# Print all evaluation metrics
print("Accuracy:", rf_accuracy)
print("Precision:", rf_precision)
print("Recall:", rf_recall)
print("F1 Score:", rf_f1_score)
print("Confusion Matrix:\n", rf_confusion)
print("Classification Report:\n", rf_classification_report)


Accuracy: 0.9528535980148883
Precision: 0.9020979020979021
Recall: 0.9626865671641791
F1 Score: 0.9314079422382671
Confusion Matrix:
 [[765  42]
 [ 15 387]]
Classification Report:
               precision    recall  f1-score   support

           0       0.98      0.95      0.96       807
           1       0.90      0.96      0.93       402

    accuracy                           0.95      1209
   macro avg       0.94      0.96      0.95      1209
weighted avg       0.95      0.95      0.95      1209



#The Random Forest Classifier achieved an accuracy of approximately 95.37% on the test data.

#Overall, all five models showed good predictive abilities, with Logistic Regression and Random Forest performing slightly better in terms of overall accuracy.